channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::BTreeMap;
+
+use actix_web::{http, HttpRequest};
+use maxminddb::{self, geoip2::City};
+
+use session::WsChannelSessionState;
+
+// Sender meta data, drawn from the HTTP Headers of the connection counterpart.
+#[derive(Serialize, Debug, Clone)]
+pub struct SenderData {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub ua: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub addr: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub city: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub region: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub country: Option<String>,
+}
+
+impl Default for SenderData {
+    fn default() -> Self {
+        Self {
+            ua: None,
+            addr: None,
+            city: None,
+            region: None,
+            country: None,
+        }
+    }
+}
+
+// Parse the Accept-Language header to get the list of preferred languages.
+// We default to "en" because of well-established Anglo-biases.
+fn preferred_language(alheader: String) -> Vec<String> {
```

Comment:
I'm surprised there isn't builtin functionality for this...

Anyways, how much do we care about parsing edge-cases in our handling of this header?  Will it cope OK with "*" as the language, as in the examples at [1]?

[1] https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#Syntax

Reply:
I'd forgotten about "*", thanks! Added code to deal with it.